File: src/main/java/org/nyxlang/interpreter/Interpreter.java
package org.nyxlang.interpreter;

import org.jetbrains.annotations.NotNull;
import org.nyxlang.interpreter.exception.InterpreterException;
import org.nyxlang.interpreter.exception.VisitorException;
import org.nyxlang.interpreter.memory.ActivationRecord;
import org.nyxlang.interpreter.memory.IActivationRecord;
import org.nyxlang.interpreter.visitor.*;
import org.nyxlang.parser.ast.INode;
import org.nyxlang.symbol.ISymbolTable;
import org.nyxlang.symbol.SymbolTable;

import java.util.ArrayList;
import java.util.List;

public class Interpreter implements IInterpreter {

    private IActivationRecord globalMemory = new ActivationRecord(null);
    private ISymbolTable globalSymbolTable = new SymbolTable(null);
    private List<IVisitor> visitorsList = new ArrayList<>(32) {
        {
            add(new ProgramVisitor());
            add(new UnaryOperationVisitor());
            add(new BinaryOperationVisitor());
            add(new BoolVisitor());
            add(new NumberVisitor());
            add(new IfVisitor());
            add(new LoopVisitor());
            add(new NativeVisitor());
            add(new StatementListVisitor());
            add(new VarAccessVisitor());
            add(new VarAssignVisitor());
            add(new VarDeclareVisitor());
            add(new TypeVisitor());
        }
    };

    @NotNull
    @Override
    public Object interpret(@NotNull INode ast) {
        return interpret(ast, null);
    }

    @NotNull
    @Override
    public Object interpret(@NotNull INode ast, ISymbolTable symbolTable) {
        if (symbolTable != null) {
            globalSymbolTable = symbolTable;
        }
        return evalNode(ast);
    }

    @Override
    public Object evalNode(INode node) {
        if (node == null) {
            return null;
        }

        List<InterpreterError> errors = new ArrayList<>(32);

        for (IVisitor visitor : visitorsList) {
            if (!visitor.matches(node)) {
                continue;
            }
            try {
                return visitor.visit(this, node);
            } catch (VisitorException e) {
                errors.add(new InterpreterError(e.getMessage()));
            }
        }

        if (errors.size() > 0) {
            throw new InterpreterException("The interpreter detected an error during runtime", errors);
        }

        return null;
    }

    @NotNull
    @Override
    public IActivationRecord getGlobalMemory() {
        return globalMemory;
    }

    @NotNull
    @Override
    public ISymbolTable getSymbolTable() {
        return globalSymbolTable;
    }
}

File: src/main/java/org/nyxlang/interpreter/Interpreter.kt
package org.nyxlang.interpreter

import org.nyxlang.interpreter.exception.DynamicSemanticException
import org.nyxlang.interpreter.exception.VisitorException
import org.nyxlang.interpreter.memory.ActivationRecord
import org.nyxlang.interpreter.memory.CallStack
import org.nyxlang.interpreter.result.IRuntimeResult
import org.nyxlang.interpreter.result.ListRuntimeResult
import org.nyxlang.interpreter.result.RuntimeResult
import org.nyxlang.interpreter.result.emptyResult
import org.nyxlang.interpreter.visitor.*
import org.nyxlang.parser.ast.INode

/**
 * Implementation of the interpreter specification.
 */
class Interpreter : IInterpreter {

    private val visitorsList = arrayListOf(
            ProgramVisitor(),
            BlockVisitor(),
            UnaryOperationVisitor(),
            BinaryOperationVisitor(),
            BoolVisitor(),
            NumberVisitor(),
            IfVisitor(),
            WhenVisitor(),
            LoopVisitor(),
            BreakVisitor(),
            ContinueVisitor(),
            StatementListVisitor(),
            VarAccessVisitor(),
            VarAssignVisitor(),
            VarDeclareVisitor(),
            FunCallVisitor(),
            FunDeclareVisitor(),
            TypeVisitor())

    override val callStack = CallStack()
    override fun interpret(ast: INode) = unrollResult(evalNode(ast))

    init {
        callStack.push(ActivationRecord(name = "global"))
    }

    override fun evalNode(node: INode?): IRuntimeResult {
        if (node == null) {
            return emptyResult()
        }
        val errors = arrayListOf<InterpreterError>()
        for (visitor in visitorsList) {
            if (!visitor.matches(node)) continue
            try {
                return visitor.visit(this, node)
            } catch (e: VisitorException) {
                errors.add(InterpreterError(e.message!!))
            }
        }
        if (errors.size > 0) {
            throw DynamicSemanticException("The interpreter detected an error during runtime", errors)
        }
        return emptyResult()
    }

    fun unrollResult(result: Any): List<Any> {
        if (result is ListRuntimeResult) {
            return result.data.flatMap { unrollResult(it) }
        }
        if (result is RuntimeResult && result.data != null) {
            return listOf(result.data!!)
        }
        return listOf()
    }
}